Update community form

barbarah opened the pull request:
Based on a new design, the pull request contains the following:

- Update the community form with the licence and remove the slug. There are also descriptions added to the licence and ORCID iD; I'm still waiting for the final text.
- Reorganize the buttons on the community page for a better user experience. Instead of one button-group, move the 'manage member' button to the member list and the 'more settings' to the form.
- Show a warning message if a community lacks an ORCID iD or licence

I also fixed a caching bug I knew existed for a while but could not reproduce. I finally found the problem. When updating the community with exactly the same values I have used in the past, I would get a 200, but nothing changed. The same happened when I clicked the 'join community' button twice after leaving the community in between. The action responded with a 200, but I did not join the community. Somehow, a cached response is returned without performing the actions. I tried using all the [Route Segment options](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config), but nothing worked.

The solution to this issue seems to work, but it is a bit weird: https://github.com/vercel/next.js/discussions/50045#discussioncomment-7218266

Luckily, Next.js 14 has a better solution with [`noStore()`](https://nextjs.org/docs/app/api-reference/functions/unstable_noStore). I will use that after updating.

barbarah commented:
I'm putting this review on hold. This review contains a reorganisation of community logic. For the next feature (adding user enrichment) I'm reorganising it a bit differently. I don't want @sdevalk to need to review two versions of the same code reorganization. I will update this pull request soon with the final version.

barbarah commented:
I have updated the community logic. Now there are functions for server actions are in `community/actions.ts`, hooks only for client code in `community/hooks.tsx`, the converters that can be used for both the client and the server can be found in `community/clerk-converters` and the definitions (client/server) can be found in `community/defenitions.ts`.

barbarah commented on apps/researcher/src/lib/community/definitions.ts:
```diff
@@ -0,0 +1,28 @@
+  imageUrl: string;
+  createdAt: number;
+  membershipCount?: number;
+  licence?: string;
```
Question for @sdevalk: I chose the term `licence`. (British English) I see the enricher API uses `license` (American English). Maybe we should choose one, are we using British or American English?

sdevalk replied:
For the data model of the 'backend of the frontend' I use elements from the Schema.org vocabulary. This vocab uses `en-us`, e.g. [license](https://schema.org/license). I prefer to use this style because most models and vocabs that I know of use it. However, for the display to the user we can use `en-gb`.

barbarah replied:
I have renamed it to `license`. I see Gertjan also uses `license` in the text he provided. I will keep it the way Gertjan wrote it.

sdevalk commented on apps/researcher/.env.development:
```diff
@@ -5,3 +5,5 @@
 SPARQL_ENDPOINT_URL=https://api.colonial-heritage.triply.cc/datasets/data-hub-testing/knowledge-graph/services/kg/sparql
 DATASET_BROWSER_URL=https://dev.datasets.colonialcollections.nl
+
+COMMUNITY_LICENCE="https://creativecommons.org/licenses/by/4.0/deed.en"
```
Should we make the scope of the license clear - that it's about enrichments provided by users?

barbarah replied:
Renamed to `COMMUNITY_ENRICHMENT_LICENSE`

sdevalk commented on apps/researcher/src/app/[locale]/(objects)/objects/[id]/object-lists-menu.tsx:
```diff
@@ -123,14 +124,12 @@ export default function ObjectListsMenu({objectId}: ObjectListsMenuProps) {
   const {user} = useUser();
 
-  if (!user || !user.organizationMemberships.length) {
+  const communities = useUserCommunities();
```
For consistency, rename `communities` to `userCommunities` or rename `useUserCommunities` to `useCommunities`?

sdevalk commented on apps/researcher/src/app/[locale]/communities/[slug]/buttons.tsx:
```diff
@@ -77,86 +74,20 @@ export function JoinCommunityButton({communityId, communitySlug}: Props) {
-        },
-      },
-    });
+  const manageMembersClick = () => {
```
Would the condensed version also work, `const manageMembersClick = () => openProfile(...)`?

sdevalk commented on apps/researcher/src/app/[locale]/communities/[slug]/edit-community-form.tsx:
```diff
@@ -16,25 +17,22 @@ interface Props {
-  slug: string;
   description: string;
   attributionId: string;
+  licenceChecked: boolean;
```
Does a user 'check' a license? Or 'accept' or 'agree to' a license?

barbarah replied:
I have renamed the property to `agreedToLicense`

sdevalk commented on apps/researcher/src/app/[locale]/communities/[slug]/edit-community-form.tsx:
```diff
@@ -16,25 +17,22 @@ interface Props {
   slug: string;
   description?: string;
   attributionId?: string;
+  licence?: string;
+  licenceToAccept: string;
```
I don't fully understand the difference between these props - to me it suggests there can be a difference between a license and a license that can be accepted. Is that correct?

barbarah replied:
I have removed `licenceToAccept` here, this is just the license in the environment variable that will be used to set the license if checked. But instead of passing the environment variable, I have imported it there where it is needed.

sdevalk commented on apps/researcher/src/app/[locale]/communities/[slug]/page.tsx:
```diff
@@ -162,6 +166,8 @@ export default async function CommunityPage({params}: Props) {
                 slug={community.slug!}
                 attributionId={community.attributionId}
+                licence={community.licence}
+                licenceToAccept={env['COMMUNITY_LICENCE'] || ''}
```
What happens if the env var does not exist?

barbarah replied:
Good point; this should be an error because we do not want to save an empty string silently for the community.

sdevalk commented on apps/researcher/src/lib/community/actions.ts:
```diff
@@ -99,6 +50,9 @@ export function sort(communities: Community[], sortBy: SortBy) {
     } else if (sortBy === SortBy.NameDesc) {
       return b.name.localeCompare(a.name);
     } else if (sortBy === SortBy.CreatedAtDesc) {
+      if (typeof a.createdAt !== 'number' || typeof b.createdAt !== 'number') {
+        throw new Error('createdAt must be of type number');
+      }
```
Is there a way to do the validation outside of the sort ('separation of concerns')?

barbarah replied:
This is not needed anymore as this is part of the Clerk converter. I will remove this check.

sdevalk commented on apps/researcher/src/lib/community/clerk-converters.test.ts:
```diff
@@ -0,0 +1,193 @@
+  });
+
+  it('converts an OrganizationResource to a Community', () => {
+    const organization: OrganizationResource = {
```
For readability: do we need to implement all props and functions of an `OrganizationResource` here or can we only provide the props that are relevant for the `organizationToCommunity` conversion?

barbarah replied:
I would love to do this, but without all the props, it is not a valid `OrganizationResource`.

sdevalk replied:
I understand. For the unit test, though, it doesn't seem to matter - `organizationToCommunity` only needs specific props from `OrganizationResource`? If so, you could silence TS by setting a `// @ts-expect-error:TS{error code}`

barbarah replied:
Yes, this will work, thank you.

sdevalk commented on apps/researcher/src/lib/community/hooks.tsx:
```diff
@@ -0,0 +1,68 @@
+    );
+
+    if (canAddEnrichments) {
+      return communities.filter(community => {
+        return community.canAddEnrichments;
+      });
```
Would the shorthand also work (similar to line 56)?

sdevalk commented on apps/researcher/src/messages/en/licence.mdx:
```diff
@@ -0,0 +1,6 @@
+- copied and redistributed in any medium or format for any purpose, even commercially.
+- adapted, remixed, transformed, and built upon for any purpose, even commercially.
+
+Users must give appropriate credit to the community and provide a link to the license. Any changes must be indicated and may not suggests that the community endorses them.
```
`suggest`?

sdevalk commented on apps/researcher/src/messages/en/messages.json:
```diff
@@ -179,22 +179,21 @@
     "labelDescription":"Description",
-    "labelAttributionId":"ORCID iD",
+    "labelAttributionId":"ORCID",
+    "descriptionAttributionId": "A community needs at least one member with a valid ORCID to comment and add new information to objects in the application. ORCID is primarily intended for researchers and can be created through the <link>ORCID-website</link>.\n\nWhen you link your ORCID to your community, you will be seen as its spokesperson and all comments made by the members are registered usin ...
```
`ORCID website` (without the hyphen)?

sdevalk commented on apps/researcher/src/messages/nl/licence.mdx:
```diff
@@ -0,0 +1,6 @@
+- mag worden gekopieerd, verspreid, of doorgegeven via elk medium of bestandsformaat voor alle doeleinden, inclusief commerciële doeleinden.
+- mag worden bewerkt, geremixt, veranderd, of dat er voor alle doeleinden, inclusief commerciële doeleinden, afgeleide werken van gemaakt mogen worden.
+
+Gebruiker moeten de community als maker van het werk vermelden , een link naar de licentie plaatsen en aangeven of het werk veranderd is. Veranderingen moeten kenbaar gemaakt worden en mogen niet de indruk wekken dat de community ermee instemt.
```
1. `Gebruikers` (similar to the English text)?
2. `vermelden, een` (remove space)

sdevalk commented on apps/researcher/src/messages/en/messages.json:
```diff
@@ -179,22 +179,21 @@
-    "noAttributionIdWarning": "To add narratives or provenance to object you need to add an ORCID iD to your community profile.",
-    "addAttributionIdButton": "Add ORCID iD"
+    "attributionIdInvalidString": "ORCID should be a valid url",
+    "noAttributionIdWarning": "To add narratives or provenance to object you need to add an ORCID to your community profile.",
```
`or provenance information to objects`?

sdevalk commented on apps/researcher/src/messages/nl/messages.json:
```diff
@@ -179,22 +179,22 @@
-    "noAttributionIdWarning": "Om verhalen of herkomst aan object toe te voegen, moet u een ORCID iD aan uw communityprofiel toevoegen.",
-    "addAttributionIdButton": "ORCID iD toevoegen"
+    "attributionIdInvalidString": "ORCID moet een geldige URL zijn",
+    "noAttributionIdWarning": "Om verhalen of herkomst aan object toe te voegen, moet u een ORCID aan uw communityprofiel toevoegen.",
```
`of herkomstinformatie aan objecten`?

barbarah commented:
I have merged the `main` branch in this pull request with Next.js 14. This means I can use the new function `unstable_noStore`. I have removed my temporary fix. I have added `unstable_noStore` to:

- Functions that perform a state change to the Clerk data: `updateCommunity` and `joinCommunity`. 
- Functions where the output changes based on the authentication state (logged in or out): `isAdmin` and `isMember.`

Maybe we need to check all our external APIs to find out where the `unstable_noStore` is required. But I will not do this in this pull request.

barbarah commented:
I renamed the environment variable to NEXT_PUBLIC_COMMUNITY_ENRICHMENT_LICENSE so it can be used directly in the client component it is used. The value will be set on build time. In Next.js client component, you cannot use `import {env} from 'node:process';`, but you can do `process.env['NEXT_PUBLIC_COMMUNITY_ENRICHMENT_LICENSE']` directly in code.

sdevalk commented on apps/researcher/src/lib/community/hooks.tsx:
```diff
@@ -0,0 +1,66 @@
+    openOrganizationProfile({
+      afterLeaveOrganizationUrl: `/revalidate/?path=/[locale]/communities/${communitySlug}&redirect=/communities/${communitySlug}`,
+      customPages: ['settings', 'members']
+        .sort((a, b) => {
```
For readability, use meaningful names? E.g. `customPageA`, `customPageB`?

sdevalk commented on apps/researcher/src/lib/community/clerk-converters.test.ts:
There are some GitHub Actions warnings in this file.

sdevalk reviewed (approved):
Nice! I've added one comment.
